text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+
+# Detailed Design
+
+## Prototype
+
+See [monocdk-experiment](https://www.npmjs.com/package/monocdk-experiment).
+
+[PR](https://github.com/awslabs/aws-delivlib/pull/245) with the migration of aws-delivlib to monocdk.
+
+## cx-api
+
+The cx-api module is used to coordinate the protocol between the CDK apps and the CLI. Today, both the CLI and the framework are dynamically linked against this module (it is defined in `dependencies`). Once we ship the CDK as a single monolithic module, we will need to decide how to coordinate the protocol.
+
+The current thinking is to *statically link* the protocol definitions between the CLI and the framework. This means that at build time we will create a copy of the protocol definitions (the `.ts` files) and use it both in the CLI and the framework.
+
+## TODO
+
+- [ ] We lose per-module analytics which means we will to move to report
+  analytics at the construct level.
+- [ ] We jsii to support the concept of "submodules". I believe this is easier
+  than arbitrary namespaces which we decided not to support in jsii. One thing to consider for example, is where do submodule README files go.
+- [ ] `constructs`
+- [ ] `@aws-cdk/assert` library. Do we have to do the assert library rewrite as a jsii module (we want to do that anyway!).
+- [ ] Rewrite all example code (see [cdk-rewrite-mono-import](https://github.com/rix0rrr/cdk-rewrite-mono-imports)).
+- [ ] Reference documentation needs to also support submodules/namespaces and use the submodule's README file.
+
+# Drawbacks
+
+- [ ] In JavaScript/TypeScript code will always be fully qualified
+  (`aws_s3.Bucket` instead of `Bucket`) due to limitations of the `import` tool.
+- [ ] The size of the single module will be large (~50MiB) which can create issues of using the CDK in non-standard environments such as from a Lambda function.
+- [ ] Breaking change
+
+# Rationale and Alternatives
+
+I believe our users think of the AWS CDK as a "standard library" (or a
+"framework") and not as another library that they depend on to build their
+applications (like the AWS SDK for example). When users create CDK libraries and
+apps they don't think of the AWS CDK as yet another library that they use, they
+think of the AWS CDK as the foundation of their app.
+
+We can draw the analogy to other standard libraries like the Node.js SDK, the
+JDK, the .NET Framework. When users write Node.js libraries or apps they don't "depend" on the Node.js SDK, they actually "peer depend" on it (it is expressed via a special `engines` section in `package.json` but for all intents and purposes it's behaves like a peer dependency).
+
+I would argue that if a vendor publishes a 3rd-party construct library, what they *really* want to say is "I am compatible with CDK >= 1.23.0". Then, the decision about which actual CDK version is being used is left to the app level.
+
+## Alternatives Considered
```

Comment:
There needs to be a section on a model that explores "not 130 packages" but not one package. There is a middle ground that could make sense. In the Angular world for instance (which they also consider to be a platform), we have a core set to make apps "work". The core, the "platform" (just the app environment, e.g. browser, server, etc), and the compiler. CDK operates much the same way, where core = constructs, platform = cfn, and compiler = also core, cx-api, etc. Angular has other packages for commonly used utilities like common (includes template extensions), forms, router, animations, etc. The Angular CLI installs all or most of these for new apps, and then manages them moving forward. There are less than ten in total, and library authors typically peer depend on a subset, again shielded by major version protection.

Updating/keeping dependencies in line is seen as relatively stress-free, in part due to the excellent tooling developed, and also just that the level of churn in the Angular world is relatively low (breaking changes don't typically impact everyone, and are very uncommon). HapiJS works in much the same way, where the core engine is exported under the main package, but utilities (very commonly used utilities) are left for the user to install and manage.

Reply:
See [discussion](https://github.com/aws/aws-cdk-rfcs/pull/122/files#r385272327). The main concern is the fact that more than a single module will require 3rd party libraries to bump a new major version every time they add a dependency, or otherwise violate semver.

I will add this to **alternatives considered** with the above reasoning.

Reply:
added